### DISCUSSION TITLE
Recommanded way to have multiple staticlib with corrosion

### DISCUSSION THREAD

**Discussion by OlivierLDff** [corrosion-rs/corrosion#401] [General]
I have 2 crate that depends on `cxx`, when I try to create a project that link to the 2 of them, I get linking error, as symbols for cxx object are defined more than once.

I've stumble across that: https://github.com/dtolnay/cxx/issues/883

I'm unsure how to solve the problem in practice. I will investigate on my side, but has anyone sucessfully integrated 2 cxx crates using corrosion? Is there an example somewhere?

**Comment by jschwe** [comment]
It's not a `cxx` issue, it's a general rust issue: https://github.com/rust-lang/rust/issues/104707

I've been thinking (but haven't had the time to implement) about adding an abstraction to Corrosion that does the following:
- Create a dummy wrapper package that has a `staticlib` library target, and depends on all the libraries you need. This should be automatically created by a CMake function `corrosion_create_wrapper_staticlib(new_lib_name dep_1_crate ...)`.
- The dependencies would need to have an `rlib` target, so that cargo can link them into the wrapper crate.
- corrosion currently ignores `rlib` only libraries, so potentially one would need to add a `staticlib` to import the actual cxx libraries, even if they will not be used directly, just to make corrosion aware that these libararies exist.

**Comment by OlivierLDff** [comment]
Ok, have you ever integrated 2 libraries yet with a workaround?
I guess I will go with the one crate with everything in it as it seems to be the faster solution now.

**Comment by jschwe** [comment]
I converted this to a discussion, to keep the issue tracker cleaner.

Related, I found this blog post on the topic helpful: https://alanwu.space/post/symbol-hygiene/

I'm planning on adding some documentation in the `common issues` chapter, but I haven't gotten to it yet.

**Comment by OlivierLDff** [comment]
I guess the solution having everything `rlib`, and corrosion create a stub `staticlib` would be the nicest for the end user to avoid all conflict.
This would even not break API, since `rlib` is not supported at the moment. So to get this new feature, simply add `CRATE_TYPE` to `rlib`.
